File: js/test_websocket.js
/**
 * Test script for WebSocket bindings
 * Tests WebSocketClient, StockWebSocketClient, FutOptWebSocketClient
 */

const { WebSocketClient, StockWebSocketClient, FutOptWebSocketClient } = require('./index.js');

console.log('=== JavaScript WebSocket Bindings Test ===\n');

// Test 1: Module loading
console.log('Test 1: Module loading');
try {
  console.log('  - WebSocketClient:', typeof WebSocketClient);
  console.log('  - StockWebSocketClient:', typeof StockWebSocketClient);
  console.log('  - FutOptWebSocketClient:', typeof FutOptWebSocketClient);
  console.log('  Result: OK\n');
} catch (e) {
  console.log('  FAILED:', e.message, '\n');
  process.exit(1);
}

// Test 2: WebSocketClient creation
console.log('Test 2: WebSocketClient creation');
try {
  const ws = new WebSocketClient('test-api-key');
  console.log('  - Created WebSocketClient:', ws !== null);
  console.log('  Result: OK\n');
} catch (e) {
  console.log('  FAILED:', e.message, '\n');
  process.exit(1);
}

// Test 3: Stock/FutOpt client accessors
console.log('Test 3: Stock/FutOpt client accessors');
try {
  const ws = new WebSocketClient('test-api-key');
  const stock = ws.stock;
  const futopt = ws.futopt;

  console.log('  - ws.stock exists:', stock !== undefined);
  console.log('  - ws.stock.on:', typeof stock.on);
  console.log('  - ws.stock.connect:', typeof stock.connect);
  console.log('  - ws.stock.subscribe:', typeof stock.subscribe);
  console.log('  - ws.stock.unsubscribe:', typeof stock.unsubscribe);
  console.log('  - ws.stock.disconnect:', typeof stock.disconnect);
  console.log('  - ws.stock.isConnected:', typeof stock.isConnected);

  console.log('  - ws.futopt exists:', futopt !== undefined);
  console.log('  - ws.futopt.on:', typeof futopt.on);
  console.log('  - ws.futopt.connect:', typeof futopt.connect);
  console.log('  - ws.futopt.subscribe:', typeof futopt.subscribe);
  console.log('  Result: OK\n');
} catch (e) {
  console.log('  FAILED:', e.message, '\n');
  process.exit(1);
}

// Test 4: Event handler registration
console.log('Test 4: Event handler registration');
try {
  const ws = new WebSocketClient('test-api-key');

  // Register handlers
  ws.stock.on('message', (data) => console.log('  [stock] message:', data));
  ws.stock.on('connect', () => console.log('  [stock] connected'));
  ws.stock.on('disconnect', (reason) => console.log('  [stock] disconnected:', reason));
  ws.stock.on('error', (err) => console.log('  [stock] error:', err));

  ws.futopt.on('message', (data) => console.log('  [futopt] message:', data));
  ws.futopt.on('connect', () => console.log('  [futopt] connected'));
  ws.futopt.on('disconnect', (reason) => console.log('  [futopt] disconnected:', reason));
  ws.futopt.on('error', (err) => console.log('  [futopt] error:', err));

  console.log('  - All event handlers registered');
  console.log('  Result: OK\n');
} catch (e) {
  console.log('  FAILED:', e.message, '\n');
  process.exit(1);
}

// Test 5: Invalid event type
console.log('Test 5: Invalid event type handling');
try {
  const ws = new WebSocketClient('test-api-key');

  try {
    ws.stock.on('invalid_event', () => {});
    console.log('  FAILED: Should have thrown error\n');
    process.exit(1);
  } catch (e) {
    if (e.message.includes('Unknown event type')) {
      console.log('  - Correctly rejected invalid event type');
      console.log('  Result: OK\n');
    } else {
      console.log('  FAILED: Unexpected error:', e.message, '\n');
      process.exit(1);
    }
  }
} catch (e) {
  console.log('  FAILED:', e.message, '\n');
  process.exit(1);
}

// Test 6: isConnected getter
console.log('Test 6: isConnected getter');
try {
  const ws = new WebSocketClient('test-api-key');

  const stockConnected = ws.stock.isConnected;
  const futoptConnected = ws.futopt.isConnected;

  console.log('  - ws.stock.isConnected (initial):', stockConnected);
  console.log('  - ws.futopt.isConnected (initial):', futoptConnected);

  if (stockConnected === false && futoptConnected === false) {
    console.log('  Result: OK\n');
  } else {
    console.log('  FAILED: Expected false for isConnected\n');
    process.exit(1);
  }
} catch (e) {
  console.log('  FAILED:', e.message, '\n');
  process.exit(1);
}

// Test 7: Subscribe validation (without connection)
console.log('Test 7: Subscribe validation');
try {
  const ws = new WebSocketClient('test-api-key');

  // Should fail because not connected
  try {
    ws.stock.subscribe({ channel: 'trades', symbol: '2330' });
    console.log('  FAILED: Should have thrown error (not connected)\n');
    process.exit(1);
  } catch (e) {
    if (e.message.includes('Not connected')) {
      console.log('  - Correctly rejected subscribe before connect');
      console.log('  Result: OK\n');
    } else {
      console.log('  FAILED: Unexpected error:', e.message, '\n');
      process.exit(1);
    }
  }
} catch (e) {
  console.log('  FAILED:', e.message, '\n');
  process.exit(1);
}

// Test 8: TypeScript definitions check
console.log('Test 8: TypeScript definitions');
try {
  const fs = require('fs');
  const dts = fs.readFileSync('./index.d.ts', 'utf8');

  const hasWebSocketClient = dts.includes('export declare class WebSocketClient');
  const hasStockWebSocketClient = dts.includes('export declare class StockWebSocketClient');
  const hasFutOptWebSocketClient = dts.includes('export declare class FutOptWebSocketClient');
  const hasOnMethod = dts.includes('on(event: string');
  const hasConnectMethod = dts.includes('connect()');
  const hasSubscribeMethod = dts.includes('subscribe(options:');
  const hasIsConnected = dts.includes('get isConnected()');

  console.log('  - WebSocketClient class:', hasWebSocketClient);
  console.log('  - StockWebSocketClient class:', hasStockWebSocketClient);
  console.log('  - FutOptWebSocketClient class:', hasFutOptWebSocketClient);
  console.log('  - on() method:', hasOnMethod);
  console.log('  - connect() method:', hasConnectMethod);
  console.log('  - subscribe() method:', hasSubscribeMethod);
  console.log('  - isConnected getter:', hasIsConnected);

  if (hasWebSocketClient && hasStockWebSocketClient && hasFutOptWebSocketClient &&
      hasOnMethod && hasConnectMethod && hasSubscribeMethod && hasIsConnected) {
    console.log('  Result: All OK\n');
  } else {
    console.log('  FAILED: Missing some TypeScript definitions\n');
    process.exit(1);
  }
} catch (e) {
  console.log('  FAILED:', e.message, '\n');
  process.exit(1);
}

console.log('=== All Tests Passed ===\n');
console.log('Note: Connection tests require valid API key and network.');
console.log('Usage with real API key:');
console.log('  const ws = new WebSocketClient("your-api-key");');
console.log('  ws.stock.on("message", (data) => console.log(JSON.parse(data)));');
console.log('  ws.stock.on("connect", () => {');
console.log('    ws.stock.subscribe({ channel: "trades", symbol: "2330" });');
console.log('  });');
console.log('  ws.stock.connect();');

File: py/test_websocket.py
#!/usr/bin/env python3
"""
Test script for WebSocket bindings.

Tests:
1. WebSocketClient creation and attribute access
2. Callback registration (on/off)
3. StockWebSocketClient methods
4. FutOptWebSocketClient methods
5. MessageIterator creation
6. Basic GC check

Note: Actual connection tests require a valid API key and network access.
"""

import gc
import sys


def test_import():
    """Test importing the module."""
    print("Test 1: Import module...")
    import fugle_marketdata
    print(f"  Module: {fugle_marketdata}")
    print(f"  WebSocketClient: {fugle_marketdata.WebSocketClient}")
    print(f"  StockWebSocketClient: {fugle_marketdata.StockWebSocketClient}")
    print(f"  FutOptWebSocketClient: {fugle_marketdata.FutOptWebSocketClient}")
    print(f"  MessageIterator: {fugle_marketdata.MessageIterator}")
    print("  OK\n")
    return fugle_marketdata


def test_client_creation(fugle_marketdata):
    """Test WebSocket client creation."""
    print("Test 2: Create WebSocketClient...")
    ws = fugle_marketdata.WebSocketClient("test-api-key")
    print(f"  Client: {ws}")

    # Access stock and futopt getters
    stock = ws.stock
    print(f"  Stock client: {stock}")

    futopt = ws.futopt
    print(f"  FutOpt client: {futopt}")
    print("  OK\n")
    return ws


def test_callback_registration(fugle_marketdata):
    """Test callback registration."""
    print("Test 3: Test callback registration...")
    ws = fugle_marketdata.WebSocketClient("test-api-key")

    # Register callbacks
    messages_received = []

    def on_message(msg):
        messages_received.append(msg)

    def on_connect():
        print("    Connected!")

    def on_disconnect(code, reason):
        print(f"    Disconnected: {code} - {reason}")

    def on_error(message, code):
        print(f"    Error: {code} - {message}")

    # Register on stock client
    stock = ws.stock
    stock.on("message", on_message)
    stock.on("connect", on_connect)
    stock.on("disconnect", on_disconnect)
    stock.on("error", on_error)
    print("  Callbacks registered")

    # Test off (unregister)
    stock.off("message")
    print("  Message callback unregistered")

    # Test invalid event type
    try:
        stock.on("invalid_event", on_message)
        print("  ERROR: Should have raised ValueError")
        return False
    except ValueError as e:
        print(f"  Invalid event correctly rejected: {e}")

    # Test non-callable
    try:
        stock.on("message", "not a function")
        print("  ERROR: Should have raised TypeError")
        return False
    except TypeError as e:
        print(f"  Non-callable correctly rejected: {e}")

    print("  OK\n")
    return True


def test_not_connected_error(fugle_marketdata):
    """Test error when not connected."""
    print("Test 4: Test not connected errors...")
    ws = fugle_marketdata.WebSocketClient("test-api-key")
    stock = ws.stock

    # is_connected should return False
    assert not stock.is_connected(), "Should not be connected"
    print("  is_connected() returns False")

    # subscribe should fail when not connected
    try:
        stock.subscribe("trades", "2330")
        print("  ERROR: Should have raised RuntimeError")
        return False
    except RuntimeError as e:
        print(f"  Subscribe correctly fails when not connected: {e}")

    # unsubscribe should fail when not connected
    try:
        stock.unsubscribe("test-id")
        print("  ERROR: Should have raised RuntimeError")
        return False
    except RuntimeError as e:
        print(f"  Unsubscribe correctly fails when not connected: {e}")

    # messages() should fail when not connected
    try:
        _ = stock.messages()
        print("  ERROR: Should have raised RuntimeError")
        return False
    except RuntimeError as e:
        print(f"  Messages correctly fails when not connected: {e}")

    print("  OK\n")
    return True


def test_futopt_client(fugle_marketdata):
    """Test FutOpt client."""
    print("Test 5: Test FutOpt client...")
    ws = fugle_marketdata.WebSocketClient("test-api-key")
    futopt = ws.futopt

    # Verify it's the correct type
    assert isinstance(futopt, fugle_marketdata.FutOptWebSocketClient)
    print(f"  FutOpt client type: {type(futopt)}")

    # Test callback registration
    futopt.on("message", lambda msg: None)
    futopt.off("message")
    print("  Callbacks work")

    # Test is_connected
    assert not futopt.is_connected()
    print("  is_connected() returns False")

    # Test invalid channel should fail during subscribe
    # (but we can't test this without connect)

    print("  OK\n")
    return True


def test_gc_check(fugle_marketdata):
    """Basic GC check - ensure no obvious memory leaks."""
    print("Test 6: Basic GC check...")

    # Create and destroy multiple clients
    for i in range(10):
        ws = fugle_marketdata.WebSocketClient("test-api-key")
        stock = ws.stock
        futopt = ws.futopt

        # Register callbacks
        stock.on("message", lambda msg: msg)
        stock.on("connect", lambda: None)
        futopt.on("error", lambda m, c: None)

        # Unregister
        stock.off("message")
        stock.off("connect")
        futopt.off("error")

        # Delete references
        del stock
        del futopt
        del ws

    # Force GC
    gc.collect()
    gc.collect()
    gc.collect()

    print(f"  GC stats: {gc.get_stats()}")
    print("  No crashes during GC - OK\n")
    return True


def test_subscriptions_list(fugle_marketdata):
    """Test subscriptions() returns empty list when not connected."""
    print("Test 7: Test subscriptions list...")
    ws = fugle_marketdata.WebSocketClient("test-api-key")
    stock = ws.stock
    futopt = ws.futopt

    # Should return empty list
    subs = stock.subscriptions()
    assert subs == [], f"Expected empty list, got {subs}"
    print(f"  Stock subscriptions: {subs}")

    subs = futopt.subscriptions()
    assert subs == [], f"Expected empty list, got {subs}"
    print(f"  FutOpt subscriptions: {subs}")

    print("  OK\n")
    return True


def main():
    """Run all tests."""
    print("=" * 60)
    print("WebSocket Bindings Test Suite")
    print("=" * 60 + "\n")

    try:
        fugle_marketdata = test_import()
        test_client_creation(fugle_marketdata)

        if not test_callback_registration(fugle_marketdata):
            return 1

        if not test_not_connected_error(fugle_marketdata):
            return 1

        if not test_futopt_client(fugle_marketdata):
            return 1

        if not test_gc_check(fugle_marketdata):
            return 1

        if not test_subscriptions_list(fugle_marketdata):
            return 1

        print("=" * 60)
        print("All tests passed!")
        print("=" * 60)
        return 0

    except Exception as e:
        print(f"\nERROR: {type(e).__name__}: {e}")
        import traceback
        traceback.print_exc()
        return 1


if __name__ == "__main__":
    sys.exit(main())
